agent/agentcontainers/api.go

```diff
@@ -1146,12 +1146,30 @@ func (api *API) maybeInjectSubAgentIntoContainerLocked(ctx context.Context, dc c
 			codersdk.DisplayAppPortForward:    true,
 		}
 
+		// We need to read the config twice. This first time is so that we can extract the
+		// agent's name.
+		if config, err := api.dccli.ReadConfig(ctx, dc.WorkspaceFolder, dc.ConfigPath, []string{}); err != nil {
+			api.logger.Error(ctx, "unable to read devcontainer config", slog.Error(err))
+		} else {
+			// NOTE(DanielleMaywood):
+			// We only want to take an agent name specified in the root customization layer.
+			// This restricts the ability for a feature to specify the agent name. We may revisit
+			// this in the future, but for now we want to restrict this behavior.
+			if name := config.Configuration.Customizations.Coder.Name; name != "" {
+				// We only want to pick this name if it is a valid name.
+				if provisioner.AgentNameRegex.Match([]byte(name)) {
+					subAgentConfig.Name = name
+				} else {
+					logger.Warn(ctx, "invalid agent name in devcontainer customization, ignoring", slog.F("name", name))
```

Comment:
```suggestion
					logger.Warn(ctx, "invalid name in devcontainer customization, ignoring", slog.F("name", name))
```
I'm thinking we should let this name affect both dev container and agent. It might get a bit confusing otherwise, or we'll need to be more specific and call this `agentName` (which seems like a bad idea to me 😅).